include/unifex/any_sender_of.hpp

```diff
@@ -0,0 +1,103 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/any_unique.hpp>
+#include <unifex/sender_concepts.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+template <typename... Values>
+using any_receiver_of =
+    any_unique_t<
+        overload<void(this_&&, Values...)>(set_value),
+        overload<void(this_&&, std::exception_ptr) noexcept>(set_error),
+        overload<void(this_&&) noexcept>(set_done)>;
+
+using any_operation_state =
+    any_unique_t<
+        overload<void(this_&) noexcept>(start)>;
+
+namespace _any_sender {
+
+// For in-place constructing non-movable operation states.
+// Relies on C++17's guaranteed copy elision.
+template <typename Fun>
+struct rvo {
+  Fun fun_;
+  operator callable_result_t<Fun>() && {
+    return ((Fun&&) fun_)();
+  }
+};
+
+template <typename Fun>
+rvo(Fun) -> rvo<Fun>;
+
+template <typename... Values>
+struct _any_connect_fn {
+  using type_erased_signature_t =
+      any_operation_state(this_&&, any_receiver_of<Values...>&&);
+
+  UNIFEX_TEMPLATE (typename Sender)
+      (requires sender_to<Sender, any_receiver_of<Values...>>)
+  friend any_operation_state
+  tag_invoke(_any_connect_fn, Sender&& s, any_receiver_of<Values...>&& r) {
+    return any_operation_state{
+      std::in_place_type<connect_result_t<Sender, any_receiver_of<Values...>>>,
+      rvo{[&]() { return connect((Sender&&) s, std::move(r)); }}
+    };
+  }
+
+  UNIFEX_TEMPLATE (typename Self)
+      (requires tag_invocable<_any_connect_fn, Self, any_receiver_of<Values...>>)
+  any_operation_state operator()(Self&& s, any_receiver_of<Values...>&& r) const {
+    return tag_invoke(*this, (Self&&) s, std::move(r));
+  }
+};
+
+template <typename... Values>
+inline constexpr _any_connect_fn<Values...> _any_connect{};
+
+template <typename... Values>
+using _any_sender_of = any_unique_t<_any_connect<Values...>>;
+
+} // namespace _any_sender
+
+template <typename... Values>
+class any_sender_of : private _any_sender::_any_sender_of<Values...> {
+  using _any_sender_of = _any_sender::_any_sender_of<Values...>;
+
+public:
+  template <template <class...> class Variant, template <class...> class Tuple>
+  using value_types = Variant<Tuple<Values...>>;
+
+  template <template <class...> class Variant>
+  using error_types = Variant<std::exception_ptr>;
+
+  using _any_sender_of::_any_sender_of;
+
+  any_operation_state connect(any_receiver_of<Values...> receiver) && {
```

Comment:
I think we can possibly get rid of the need to heap-allocate the type-erased any_receiver_of here.

The idea would be to have the connect() method return an `any_operation_state_for<Receiver>` that stores the passed receiver object and then pass an `any_receiver_ref` that holds a type-erased pointer to this receiver.

This would require a few tweaks here, but would reduce it down to two heap-allocations (one for the sender and one for the operation-state, rather than three heap-allocations.

For a similar technique, see the `type_erased_stream` implementation which uses this to type-erase the receiver.

This customistaion would also provide an opportunity for hooking up the incoming receiver's stop_token up to an inplace_stop_token to be passed down to the implementation. eg. using the `inplace_stop_token_adapter<T>` to do the adaption.

Reply:
@lewissbaker I've implemented the `any_receiver_ref` idea, but I haven't yet done the `inplace_stop_token` hookup. I may need your guidance here, but give me a chance to play with it first.